File: leetcode/algorithms/771_jewels_and_stones/main.c
int numJewelsInStones(char* jewels, char* stones) {
    int numsArray[52] = {0};
    for (int i = 0; jewels[i] != '\0'; i++) {
        if (jewels[i] > 'Z') {
            numsArray[jewels[i] - 'a' + 26] = 1;
        } else {
            numsArray[jewels[i] - 'A'] = 1;
        }
    }

    int count = 0;
    for (int i = 0; stones[i] != '\0'; i++) {
        if (stones[i] > 'Z') {
            count += numsArray[stones[i] - 'a' + 26];
        } else {
            count += numsArray[stones[i] - 'A'];
        }
    }

    return count;
}


// Solution
int solution(char* jewels, char* stones) {
    char t[128] = {};
    for (char* p = jewels; *p; ++p) {
        t[*p] = (char)1;
    }
    
    int a = 0;
    for (char* p = stones; *p; ++p) {
        a = a + t[*p];
    }

    return a;
}

File: leetcode/algorithms/771_jewels_and_stones/main.py
from typing import List


class JewelsAndStones:
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        numsArray: List[int] = [0] * 128
        for jewel in jewels:
            numsArray[ord(jewel)] = 1

        count: int = 0
        for stone in stones:
            count += numsArray[ord(stone)]

        return count


    # Solution
    # Solution 1
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        return sum(map(jewels.count, stones))

    # Solution 2
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        return sum(map(stones.count, jewels))

    # Solution 3
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        return sum(stone in jewels for stone in stones)

    # Solution 4: set
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        jewelsSet = set(jewels)
        return sum(stone in jewelsSet for stone in stones)
